provider/datagen/src/baked_exporter.rs

```diff
@@ -330,80 +334,95 @@ impl DataExporter for BakedExporter {
+                ))?;
+            }
+            let (bake, locale) = values.into_iter().next().unwrap();
+            let locale = locale.into_iter().next().unwrap();
 
-                if locale == "und" {
-                    quote! {
-                        if locale.is_empty() {
-                            Ok(Self::#singleton_ident)
-                        } else {
-                            Err(icu_provider::DataErrorKind::ExtraneousLocale)
-                        }
-                    }
-                } else if icu_locid::Locale::try_from_bytes_with_single_variant_single_keyword_unicode_extension(locale.as_bytes()).is_ok() {
-                    self.dependencies.insert("icu_locid");
-                    quote! {
-                        if icu_provider::DataLocale::from(icu_locid::locale!(#locale)).eq(&locale) {
-                            Ok(Self::#singleton_ident)
-                        } else {
-                            Err(icu_provider::DataErrorKind::MissingLocale)
-                        }
-                    }
+            if locale != "und" {
+                Err(DataError::custom("Singleton mode needs the default locale"))?;
+            }
+
+            let singleton_ident = format!("SINGLETON_{}", ident.to_ascii_uppercase())
+                .parse::<TokenStream>()
+                .unwrap();
+
+            // Exposing singleton structs separately allows us to get rid of fallibility by using
+            // the struct directly.
+            singleton = Some(quote! {
+                #[clippy::msrv = "1.61"]
+                impl $provider {
+                    #[doc(hidden)]
+                    pub const #singleton_ident: &'static #struct_type = &#bake;
+                }
+            })
+            .into_iter();
+
+            quote! {
+                if locale.is_empty() {
+                    Ok(Self::#singleton_ident)
                 } else {
-                    quote! {
-                        if locale.strict_cmp(#locale.as_bytes()).is_eq() {
-                            Ok(Self::#singleton_ident)
-                        } else {
-                            Err(icu_provider::DataErrorKind::MissingLocale)
-                        }
-                    }
+                    Err(icu_provider::DataErrorKind::ExtraneousLocale)
                 }
             }
-            _ => {
-                let mut map = BTreeMap::new();
-                let mut statics = Vec::new();
-
-                for (bake, locales) in values {
-                    let first_locale = locales.iter().next().unwrap();
-                    let anchor = syn::parse_str::<syn::Ident>(
-                        &first_locale.to_ascii_uppercase().replace('-', "_"),
-                    )
-                    .unwrap();
-                    statics.push(quote! { static #anchor: #struct_type = #bake; });
-                    map.extend(locales.into_iter().map(|l| (l, anchor.clone())));
-                }
+        } else {
```

Comment:
question: we should also be coalescing keys based on fallback here, yes? Or is that a planned followup?

Reply:
huh?

Reply:
Assuming you mean coalescing data, this happens in `lib.rs` instead of in every exporter separately

Reply:
Yes, but that just dedups values, not lookup entries, yes? From the original issue

> even though they all fall back to `sr`. BlobDataProvider and BakedDataProvider deduplicate, so this doesn't lead to data duplication, but it does mean there are nine entries in the lookup array when there only needs to be one (when fallback is enabled).

(I'm fine not fixing that here, I just wanted to be sure we _weren't_ attempting to fix that yet)

Reply:
The provider itself dedupes values, because unrelated (under fallback) locales could still have the same values.

The provider exporter logic dedupes values and keys under fallback chains (todo).

Reply:
Yeah I was mostly hoping the lookup table could be further shrunk if you know there are entries that do not need to be hit.